### PR TITLE
Fix #13503 by adding a "catch all" clause in `pickleDef` pattern match

### LINKED ISSUE #13503: anonymous class crash pickler via transparent inline cascade of a literal type member
## Compiler version

v3.1.0-RC1

## Minimized code

```Scala
trait First {type Out}
given First with {type Out = 123}

trait Second {type Out} 
transparent inline given (using f: First): Second = new Second {type Out = f.Out}

val s = summon[Second]
```

## Output (click arrow to expand)
<details>

```scala
scala.MatchError: Literal(Constant(123)) (of class dotty.tools.dotc.ast.Trees$Literal)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleDef$$anonfun$1(TreePickler.scala:343)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleDef(TreePickler.scala:347)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:573)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleStats$$anonfun$2(TreePickler.scala:371)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleStats(TreePickler.scala:371)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$26(TreePickler.scala:597)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:598)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleDef$$anonfun$1(TreePickler.scala:342)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleDef(TreePickler.scala:347)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:573)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$40$$anonfun$1(TreePickler.scala:480)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$9(TreePickler.scala:480)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:480)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$19(TreePickler.scala:524)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:530)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$19(TreePickler.scala:524)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:530)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$19(TreePickler.scala:524)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:530)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTreeUnlessEmpty(TreePickler.scala:330)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleDef$$anonfun$1(TreePickler.scala:345)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleDef(TreePickler.scala:347)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:556)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleStats$$anonfun$2(TreePickler.scala:371)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleStats(TreePickler.scala:371)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$26(TreePickler.scala:597)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:598)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleDef$$anonfun$1(TreePickler.scala:342)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleDef(TreePickler.scala:347)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:573)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleStats$$anonfun$2(TreePickler.scala:371)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleStats(TreePickler.scala:371)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$29(TreePickler.scala:613)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:69)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:613)
	at dotty.tools.dotc.core.tasty.TreePickler.pickle$$anonfun$1(TreePickler.scala:784)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.core.tasty.TreePickler.pickle(TreePickler.scala:784)
	at dotty.tools.dotc.transform.Pickler.run$$anonfun$3$$anonfun$2(Pickler.scala:69)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.transform.Pickler.run$$anonfun$1(Pickler.scala:106)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.transform.Pickler.run(Pickler.scala:106)
	at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:303)
	at scala.collection.immutable.List.map(List.scala:246)
	at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:304)
	at dotty.tools.dotc.transform.Pickler.runOn(Pickler.scala:111)
	at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:205)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
	at dotty.tools.dotc.Run.runPhases$5(Run.scala:215)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:223)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:230)
	at dotty.tools.dotc.Run.compileSources(Run.scala:166)
	at dotty.tools.dotc.Run.compile(Run.scala:150)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:39)
	at dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:88)
	at dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
	at sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:192)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:247)
	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:182)
	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:163)
	at sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
	at sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:163)
	at sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:210)
	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:528)
	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:528)
	at sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:175)
	at sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:173)
	at sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:459)
	at sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
	at sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
	at sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
	at sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:263)
	at sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:414)
	at sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:501)
	at sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:401)
	at sbt.internal.inc.Incremental$.apply(Incremental.scala:167)
	at sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:528)
	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:482)
	at sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
	at sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:420)
	at sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
	at sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2357)
	at sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2314)
	at sbt.internal.io.Retry$.apply(Retry.scala:46)
	at sbt.internal.io.Retry$.apply(Retry.scala:28)
	at sbt.internal.io.Retry$.apply(Retry.scala:23)
	at sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:31)
	at sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2310)
	at scala.Function1.$anonfun$compose$1(Function1.scala:49)
	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
	at sbt.std.Transform$$anon$4.work(Transform.scala:68)
	at sbt.Execute.$anonfun$submit$2(Execute.scala:282)
	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
	at sbt.Execute.work(Execute.scala:291)
	at sbt.Execute.$anonfun$submit$1(Execute.scala:282)
	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```
</details>

### PR DESCRIPTION
**Pull request by soronpo**

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala [hunk 1]
```diff
@@ -331,6 +331,7 @@ class TreePickler(pickler: TastyPickler) {
       tpt match {
         case _: Template | _: Hole => pickleTree(tpt)
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala:332]:
```suggestion
        case _: Template | _: Literal | _: Hole => pickleTree(tpt)
```

File: compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala [hunk 2]
```diff
@@ -331,6 +331,7 @@ class TreePickler(pickler: TastyPickler) {
...
         case _ if tpt.isType => pickleTpt(tpt)
+        case _ => //do nothing
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala:334]:
```suggestion
```

**Review by nicolasstucki — CHANGES_REQUESTED** [review]

**Comment by dwijnand** [comment]
`tests/pos/i13503.scala failed`

File: compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala [hunk 3]
```diff
@@ -329,7 +329,7 @@ class TreePickler(pickler: TastyPickler) {
       pickleName(sym.name)
       pickleParams
       tpt match {
-        case _: Template | _: Hole => pickleTree(tpt)
+        case _: Template | _: Literal | _: Hole => pickleTree(tpt)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala:332]: I think the problem is in the typer, not the pickler. `Literal` is not a TypeTree so it shouldn't show up in `tpt`, I would expect a `TypeTree(ConstantType(Constant(123)))` instead /cc @nicolasstucki

Comment (soronpo) [reply compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala:332]: So is there anything wrong with my initial commit that just ignores `Literal`? Is there really a problem in the typer?

Comment (smarter) [reply compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala:332]: well we do need to pickle the type to be able to restore it in unpickling so I don't see how the tpt field could be ignored.

Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala:332]: Yeah, your first commit fails with

```
[info] Test dotty.tools.dotc.CompilationTests.pickling started
java.lang.AssertionError: assertion failed: illegal modifier tag 111 at Addr(333), end = Addr(331) while compiling tests/pos/i13503.scala
```

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala:332]: `TypeTree(ConstantType(Constant(123)))` makes more sense. We should find where this tree is created and try to change it.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala:332]: I wondering if `Ycheck` should have caught this as it is a `TermTree` used as a type tree.

**Comment by soronpo** [comment]
From the discussion I get that this PR is the wrong approach, so I'm closing it. 
I'll be glad to submit a new PR with a little guidance of where to attack the problem.

**Comment by dwijnand** [comment]
> I'll be glad to submit a new PR with a little guidance of where to attack the problem.

There's two ways I would tackle this.

1. Find out what `mdef.symbol` id is (e.g 123), and uncomment the assert in Symbol to die when id == 123.  That will get you to where that tree is created, and maybe from there you can figure out why it isn't wrapping it in a TypeTree.

2. Alternatively you can look to adding to the -Ycheck logic, because type trees should never be a `TermTree`, so you can extend the checking.  That might give you some interesting other cases from the tests, but otherwise it's just a good check to have under Ycheck, for the future.

**Comment by smarter** [comment]
Instead of debugging using symbol id, we can use a tree `uniqueId` (via `-Yshow-tree-ids` and `-Ydebug-tree-with-id`) to figure out where the right-hand-side tree `Literal(Constant(123))` is created:

<details>

```scala
Debug tree (id=504) creation 
Literal(Constant(123))

java.lang.Exception: Stack trace
        at java.lang.Thread.dumpStack(Thread.java:1336)
        at dotty.tools.dotc.ast.Positioned.allocateId(Positioned.scala:39)
        at dotty.tools.dotc.ast.Positioned.<init>(Positioned.scala:41)
        at dotty.tools.dotc.ast.Trees$Tree.<init>(Trees.scala:56)
        at dotty.tools.dotc.ast.Trees$Literal.<init>(Trees.scala:545)
        at dotty.tools.dotc.ast.untpd$.Literal(untpd.scala:373)
        at dotty.tools.dotc.ast.tpd$.Literal(tpd.scala:57)
        at dotty.tools.dotc.ast.TypedTreeInfo.constToLiteral(TreeInfo.scala:579)
        at dotty.tools.dotc.ast.TypedTreeInfo.constToLiteral$(TreeInfo.scala:394)
        at dotty.tools.dotc.ast.tpd$.constToLiteral(tpd.scala:22)
        at dotty.tools.dotc.typer.Inliner$InlineTyper.typedSelect(Inliner.scala:1507)
        at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2725)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2817)
        at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:121)
        at dotty.tools.dotc.typer.Inliner$InlineTyper.typedUnadapted(Inliner.scala:1635)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2883)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2887)
        at dotty.tools.dotc.typer.Typer.typedType(Typer.scala:3005)
        at dotty.tools.dotc.typer.Typer.typedTypeDef(Typer.scala:2278)
        at dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$2(Typer.scala:2745)
        at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2747)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2817)
        at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:121)
        at dotty.tools.dotc.typer.Inliner$InlineTyper.typedUnadapted(Inliner.scala:1635)
[...]
```

</details>

From that we can see that the problem comes from the call to constToLiteral in https://github.com/lampepfl/dotty/blob/3a6a9499f0bd1821351f052631a0a1d3842f05c1/compiler/src/dotty/tools/dotc/typer/Inliner.scala#L1507 where `Select(Inlined(EmptyTree,List(),Ident(given_First)),Out)` (which returns true to `.isType`) gets reduced to `Literal(Constant(123))` (which returns false to `.isType`), as far as I can tell, constToLiteral isn't designed to handle type trees as input and we should add an assert to avoid misusing it like this.

**Comment by soronpo** [comment]
So if I understand correctly, we have the following issues:
* The one in #13503
* The fact that -Ycheck didn't catch this.
* The lack of assert in `constToLiteral` that could have prevented this.
Should all three be solved in the same PR or separate ones?

**Comment by smarter** [comment]
The Ycheck one is less useful I'd say, even if we add an assert there, it won't help find the root cause of any future issue, we do have assertions on tree creations that could be changed to check that tpt.isType is true (https://github.com/lampepfl/dotty/blob/3a6a9499f0bd1821351f052631a0a1d3842f05c1/compiler/src/dotty/tools/dotc/ast/Trees.scala#L822), but it's just as easy to add that by hand whenever the compiler crashes to debug it.

**Comment by soronpo** [comment]
The new PR is #13578

**Pull request closed without merging**